### DISCUSSION TITLE
Rhino bug in JSON.stringify(obj, whitelistArray)

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#4524] [Q&A]
I'm using Mirth v3.10.1 with ES6 (Rhino language version 200) but also tested with lower Mirth v3.5.2

I experienced a bug in Mirth (Rhino) with `JSON.stringify()` function when you provide an array of strings as 2nd argument that is used as a whitelist for the property names that you want to keep **and if the property names are numbers as strings**. You can see a full code example here and compare Mirth/Rhino result with any other JS engine (browser):

https://js.do/code/rhino-issue-object-keys-stringnumber-keys

Summary:
```js
var obj1 = {"2":"b","3":"c","1":"a"};
JSON.stringify(obj1, ["1","2","3"]); // returns {} in Rhino but works in JS returning {"1":"a","2":"b","3":"c"}
JSON.stringify(obj1, [1,2,3]) // returns {"1.0":"a","2":"b","3":"c"} in Rhino but works in JS returning  {"1":"a","2":"b","3":"c"}
```

I also tried in Rhino shell with the same result. It took me hours to find why Rhino returns `{}` ...

**Any idea how to tackle this?**

**Comment by tonygermano** [comment]
I'd file the the bug against Rhino https://github.com/mozilla/rhino/issues

**Reply by tom08zehn** [reply]
Thanks, did it. Found a similar reported issue and added my comments.

What happens if Rhino solves that and releases a bugfixed version? Will a next version of Mirth contain the bugfixed Rhino?

Many thanks.

**Reply by tonygermano** [reply]
They updated Rhino between mirth versions 3.9 and 3.10, but it looks like they didn't for 3.11. It should eventually make it's way back into mirth if it's fixed in Rhino.

**Reply by tonygermano** [reply]
FYI, I opened a PR with rhino to fix this. https://github.com/mozilla/rhino/pull/876

**Reply by tom08zehn** [reply]
Thanks, let's see... Still that back and forth conversion looks a bit overcomplicated to me.

I'd expect that only those properties of the object are kept in the result that _exactly_ match the items in the replacer array. I'd **not** expect that an array of integers or floats also works... but in other JS engines this works, too.

**Reply by tonygermano** [reply]
It's necessary per the javascript specification. All object keys must be a string or symbol, but internally they get optimized to Integers when possible, especially for arrays.

That's why there's no difference between
```js
a[1] = 1;
a["1"] = 2;
// a[1] == 2
```

**Comment by tonygermano** [comment]
This is now fixed in the Rhino snapshot builds. Your test cases are part of the test suite https://github.com/mozilla/rhino/blob/189b1096fced52490ddc96ea854210e7a3d7497e/testsrc/jstests/json-stringify-array-replacer.jstest

I don't know if mirth would include a snapshot build or if they will wait for the next official release. There's been quite a bit of change since Rhino 1.7.12, which is what mirth 3.10 and 3.11 use. Rhino 1.7.13 was released last September (before mirth 3.10.) This fix will be in Rhino 1.7.14. I don't know of an official release date for that, but the releases just seem to be a tag on the current state of the master branch more than a planned thing.
